src/psbt/finalizer.rs

```diff
@@ -0,0 +1,371 @@
+        match bitcoin::PublicKey::from_slice(&pk_bytes[1..script_pubkey_len - 1]) {
+            Ok(pk) => return Ok(Descriptor::Pk(pk)),
+            Err(e) => return Err(InputError::from(e)),
+        }
+    } else if script_pubkey.is_p2pkh() {
+        // 1. `Pkh`: creates a `PkH` descriptor if partial_sigs has the correponding pk
+        let partial_sig_contains_pk = inp
+            .partial_sigs
+            .iter()
+            .filter(|&(&pk, _sig)| {
+                // Indirect way to check the equivalence of pubkey-hashes.
+                // Create a pubkey hash and check if they are the same.
+                let addr = bitcoin::Address::p2pkh(&pk, bitcoin::Network::Bitcoin);
+                *script_pubkey == addr.script_pubkey()
+            })
+            .next();
+        match partial_sig_contains_pk {
+            Some((pk, _sig)) => {
+                return Ok(Descriptor::Pkh(pk.to_owned()));
+            }
+            None => {
+                return Err(InputError::MissingPubkey);
+            }
+        }
+    } else if script_pubkey.is_v0_p2wpkh() {
+        // 3. `Wpkh`: creates a `wpkh` descriptor if the partial sig has correponding pk.
+        let partial_sig_contains_pk = inp
+            .partial_sigs
+            .iter()
+            .filter(|&(&pk, _sig)| {
+                // Indirect way to check the equivalence of pubkey-hashes.
+                // Create a pubkey hash and check if they are the same.
+                let addr = bitcoin::Address::p2wpkh(&pk, bitcoin::Network::Bitcoin)
+                    .expect("Address corresponding to valid pubkey");
+                *script_pubkey == addr.script_pubkey()
+            })
+            .next();
+        match partial_sig_contains_pk {
+            Some((pk, _sig)) => {
+                return Ok(Descriptor::Wpkh(pk.to_owned()));
+            }
+            None => {
+                return Err(InputError::MissingPubkey);
+            }
+        }
+    } else if script_pubkey.is_v0_p2wsh() {
+        // 4. `Wsh`: creates a `Wsh` descriptor
+        if inp.redeem_script.is_some() {
+            return Err(InputError::NonEmptyRedeemScript);
+        }
+        if let Some(ref witness_script) = inp.witness_script {
+            if witness_script.to_v0_p2wsh() != *script_pubkey {
+                return Err(InputError::InvalidWitnessScript {
+                    witness_script: witness_script.clone(),
+                    p2wsh_expected: script_pubkey.clone(),
+                });
+            }
+            let ms = Miniscript::<bitcoin::PublicKey, Segwitv0>::parse(witness_script)?;
+            return Ok(Descriptor::Wsh(ms));
+        } else {
+            return Err(InputError::MissingWitnessScript);
+        }
+    } else if script_pubkey.is_p2sh() {
+        if inp.redeem_script.is_none() {
+            return Err(InputError::MissingRedeemScript);
+        }
+        let redeem_script = inp.redeem_script.as_ref().unwrap();
```

Comment:
This `is_none` followed by `unwrap` should be replaced with a `match` (which will add an indentation level but will eliminate a panic case)